block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```diff
@@ -62,6 +63,130 @@ final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
+          height: Long,
+          excess: Iterator[ByteString],
+          inView: Iterator[ByteString]
+      )
+      for {
+        lmMetas <- targetLatestMessages.values.toStream.traverse(this.lookupUnsafe)
+        (highestHeight, lowestHeight, lmPerValidatorHeight) = lmMetas.foldLeft(
+          (0L, Long.MaxValue, Map.empty[ByteString, Long])
+        ) {
+          case ((curHH, curLH, curLVH), meta) =>
+            val newHH  = Math.max(curHH, meta.blockNum)
+            val newLH  = Math.min(curLH, meta.blockNum)
+            val newLVH = curLVH.updated(meta.sender, meta.blockNum)
+            (newHH, newLH, newLVH)
+        }
+        // messages higher then highest latest message are outside the view
+        (highExcess, withoutHighExcess) = heightMap.partition { case (h, _) => h > highestHeight }
+
+        // messages inside range of latest message heights might contain new messages from validator that are not seen yet
+        lmRangeViews <- heightMap
+                         .range(lowestHeight, highestHeight + 1)
+                         .toStream
+                         .traverse {
+                           case (h, messages) =>
+                             for {
+                               lms <- messages.toStream
+                                       .traverse(this.lookupUnsafe)
+                                       .map(_.map(m => (m.blockHash, m.sender)))
+                               (excess, inView) = lms.partition {
+                                 case (_, sender) =>
+                                   // remove message if target view is not aware of validator,
+                                   // or latest message in full view is newer then target
+                                   val senderSeen = lmPerValidatorHeight.contains(sender)
+                                   senderSeen && h > lmPerValidatorHeight(sender)
+                               }
+                               r = HeightView(
+                                 h,
+                                 excess = excess.map { case (hash, _) => hash }.toIterator,
+                                 inView = inView.map { case (hash, _) => hash }.toIterator
+                               )
+                             } yield r
+                         }
+                         .map(_.toSet)
+        excessMessages = lmRangeViews.flatMap(_.excess) ++ highExcess.values.flatten
+
+        // check clause 3
+        fullLMMetas <- this.latestMessages
+        // filtering out genesis is required because there might be some validators bonded in genesis
+        // which are not yet in view. Its not quite pretty, but no other options.
+        notSeenSenders = fullLMMetas.filter { case (_, meta) => meta.blockNum != 0 }.keySet diff targetLatestMessages.keySet
+        lowExcess <- Stream
+                    // gather all messages of unseen validators
+                      .fromIterator(notSeenSenders.toIterator)
```

Comment:
This particular part is something that I'm not completely sure yet. This is supposed to remove messages that are not seen and are lower than lowest target latest message.